### DISCUSSION TITLE
using template string / expression in calc in an arbitrary value for transform rotate degree values

### DISCUSSION THREAD

**Discussion by haroldjeshua** [tailwindlabs/tailwindcss#12934] [Help]
i have 4 divs and i want each of it to have different rotate values on hover

i tried correct syntax using calc() in transform rotate ([src](https://stackoverflow.com/questions/41420327/use-calc-in-transformrotate)) + an arbitrary value, but on render, the template strings / expression doesn't work.

in vanilla css:
`.item:hover { transform: rotate(calc(-8deg * var(--item-index))) }`

seems to me the only div that works on hover is the first div with have a rotate value of [-8deg], the rest of the divs, doesn't rotate

![image](https://github.com/tailwindlabs/tailwindcss/assets/34917956/aecd7de2-883b-43f6-b622-59a4bc40c63e)

```jsx
<section className="grid grid-cols-2 gap-4 xs:grid-cols-4 sm:grid-cols-6">
        {items.map((item, index) => {
          const rotationDegree = (index + 1) * 8;
          return (
            <div
              key={item.id}
              className={`flex flex-col items-center justify-center text-center transition-transform hover:-translate-y-8  hover:-rotate-[${rotationDegree}deg]`}
            >
              <h1>Some text here</h1>
            </div>
          );
        })}
      </section>
```

do you guys have an idea why? thanks in advance.

**Comment by wongjn** [comment]
As per [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):

> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as _complete unbroken strings_ in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS:
>
> **Don’t construct class names dynamically**
>
> ```vue
> <div class="text-{{ error ? 'red' : 'green' }}-600"></div>
> ```
> In the example above, the strings `text-red-600` and `text-green-600` do not exist, so Tailwind will not generate those classes.
> Instead, make sure any class names you’re using exist in full:
>
> **Always use complete class names**
>
> ```vue
> <div class="{{ error ? 'text-red-600' : 'text-green-600' }}"></div>
> ```

**Comment by AM-NiceDay** [comment]
Hey @haroldjeshua!

As stated above, Tailwind CSS analyzes the source code of your project during build time and finds the classes matching certain patterns that it adds to the resulting generated CSS. It doesn't run the code and is not able to recognize runtime results. In simpler words, Tailwind CSS doesn't understand dynamic values.

In this particular case and any other case related to dynamic values, the best practice is to fallback to the `style` property in combination with CSS custom properties, which will look like this:

```jsx
/* I haven't tested the code locally, but you should get the idea. */
<section className="grid grid-cols-2 gap-4 xs:grid-cols-4 sm:grid-cols-6">
  {items.map((item, index) => {
    const rotationDegree = (index + 1) * 8;
    return (
      <div
        key={item.id}
        // set the custom property value based on the item rotation degree
        style={{ '--rotation': `${rotationDegree}deg` }}
        // apply the rotation on hover
        className={`flex flex-col items-center justify-center text-center transition-transform hover:-translate-y-8 hover:-rotate-[--rotation]`} 
      >
        <h1>Some text here</h1>
      </div>
    );
  })}
</section>
```

Alternatively, you can add the expected classes as comments, static values, or in any way that they appear as a complete string in your codebase.

Please let me know if you have any other questions.
